iis/mymodule.cpp

```diff
@@ -1114,115 +666,90 @@ CMyHttpModule::OnBeginRequest(IHttpContext* httpContext, IHttpEventProvider* pro
-
-    pRawHttpResponse->EntityChunkCount = 0;
-
-    // since we clean the APR pool at the end of OnSendRequest, we must get IIS-managed memory chunk
-    //
-    void *reqbuf = pHttpContext->AllocateRequestMemory(length);
-
-    memcpy(reqbuf, buf, length);
-
-    pDataChunk->DataChunkType = HttpDataChunkFromMemory;
-    pDataChunk->FromMemory.pBuffer = reqbuf;
-    pDataChunk->FromMemory.BufferLength = length;
+        // This could have been a simple lambda with generalized capture but C++11 doesn't support it yet.
+        // Can be replaced/simplified in C++14.
+        class BackgroundRequestProcessingTask
+        {
+        private:
+            AprRequestContext ctx;
+        public:
+            explicit BackgroundRequestProcessingTask(AprRequestContext&& ctx)
+                : ctx(std::move(ctx))
+            {}
+
+            BackgroundRequestProcessingTask(const BackgroundRequestProcessingTask&) = delete;
+            BackgroundRequestProcessingTask& operator=(const BackgroundRequestProcessingTask&) = delete;
+
+            BackgroundRequestProcessingTask(BackgroundRequestProcessingTask&& rhs)
+                : ctx(std::move(rhs.ctx))
+            {
+            }
 
-    CHAR szLength[21]; //Max length for a 64 bit int is 20
+            BackgroundRequestProcessingTask& operator=(BackgroundRequestProcessingTask&& rhs)
+            {
+                ctx = std::move(rhs.ctx);
+                return *this;
+            }
 
-    ZeroMemory(szLength, sizeof(szLength));
+            int operator()() const {
+                // We put the entire thread into background mode for the time of processing.
+                // According to https://docs.microsoft.com/en-us/windows/desktop/api/processthreadsapi/nf-processthreadsapi-setthreadpriority :
+                //
+                // The THREAD_PRIORITY_* values affect the CPU scheduling priority of the thread.
+                // For threads that perform background work such as file I/O, network I/O, or data processing,
+                // it is not sufficient to adjust the CPU scheduling priority; even an idle CPU priority thread
+                // can easily interfere with system responsiveness when it uses the disk and memory.
+                // Threads that perform background work should use the THREAD_MODE_BACKGROUND_BEGIN
+                // and THREAD_MODE_BACKGROUND_END values to adjust their resource scheduling priorities;
+                // threads that interact with the user should not use THREAD_MODE_BACKGROUND_BEGIN.
+                struct BackgroundThreadMode
+                {
+                    BackgroundThreadMode() {
+                        SetThreadPriority(GetCurrentThread(), THREAD_MODE_BACKGROUND_BEGIN);
+                    }
+                    ~BackgroundThreadMode() {
+                        SetThreadPriority(GetCurrentThread(), THREAD_MODE_BACKGROUND_END);
```

Comment:
destructor on a struct? When does it happen?

Reply:
It is being called when the object goes out of the scope where it is defined.
In this particular case, it will be called right before returning from the function.

Reply:
The reason why I prefer a scope-guard class over just putting two calls in the beginning and the end of the block is that is it not impossible that in the future some other code is added to the task that may potentially throw (after all, this whole code is C++) that will break the guarantee of ending the background mode. With RAII in place, this guarantee remains even if that occurs.